File: src/main/java/com/nayasis/simplelauncher/vo/JsonLink.java
package com.nayasis.simplelauncher.vo;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.github.nayasis.basica.base.Strings;
import com.nayasis.simplelauncher.common.Commons;
import com.nayasis.simplelauncher.jpa.entity.LinkEntity;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.ToString;

import java.time.LocalDateTime;

import static com.fasterxml.jackson.annotation.JsonCreator.Mode.DISABLED;

@Data
@ToString
@NoArgsConstructor
public class JsonLink {

    private String        title;
    private String        group;
    private String        path;
    private String        relativePath;
    private Boolean       showConsole;
    private String        option;
    private String        optionPrefix;
    private String        commandPrev;
    private String        commandNext;
    private String        description;
    private String        keyword;
    private String        icon;
    private Integer       execCount;
    private LocalDateTime lastExecDate;

    @JsonCreator( mode = DISABLED )
    public JsonLink( LinkEntity entity ) {
        this.title        = entity.getTitle();
        this.group        = entity.getGrp();
        this.path         = entity.getPath();
        this.relativePath = entity.getRelativePath();
        this.showConsole  = entity.getShowConsole();
        this.option       = entity.getOption();
        this.optionPrefix = entity.getOptionPrefix();
        this.commandPrev  = entity.getCommandPrev();
        this.commandNext  = entity.getCommandNext();
        this.description  = entity.getDescription();
        this.keyword      = entity.getKeyword();
        this.icon         = Strings.encode( entity.getIcon() );
        this.execCount    = entity.getExecCount();
        this.lastExecDate = entity.getLastExecDate();
    }

    public LinkEntity toLinkEntity() {

        LinkEntity entity = new LinkEntity();

        entity.setTitle( getTitle() );
        entity.setGrp( getGroup() );
        entity.setPath( getPath() );
        entity.setRelativePath( getRelativePath() );
        entity.setShowConsole( getShowConsole() );
        entity.setOption( getOption() );
        entity.setOptionPrefix( getOptionPrefix() );
        entity.setCommandNext( getCommandNext() );
        entity.setCommandPrev( getCommandPrev() );
        entity.setDescription( getDescription() );
        entity.setKeyword( getKeyword() );
        entity.setExecCount( getExecCount() );
        entity.setLastExecDate( getLastExecDate() );
        entity.setKeyword( Commons.getKeyword(group, keyword, description) );

        setIcon( entity );

        return entity;

    }

    public void setIcon( String icon ) {
        this.icon = icon;
    }

    private void setIcon( LinkEntity entity ) {
        try {
            entity.setIcon( (byte[]) Strings.decode( icon ) );
        } catch( ClassCastException e ) {}
    }

    public void setLastExecDate( LocalDateTime lastExecDate ) {
        this.lastExecDate = lastExecDate;
    }
}

File: src/main/kotlin/com/github/nayasis/simplelauncher/jpa/vo/JsonLink.kt
package com.github.nayasis.simplelauncher.jpa.vo

import com.github.nayasis.kotlin.basica.annotation.NoArg
import com.github.nayasis.kotlin.basica.core.string.decodeBase64
import com.github.nayasis.kotlin.basica.core.string.encodeBase64
import com.github.nayasis.simplelauncher.jpa.entity.Link
import java.time.LocalDateTime

@NoArg
data class JsonLink(
    var title: String? = null,
    var group: String? = null,
    var path: String? = null,
    var relativePath: String? = null,
    var showConsole: Boolean = false,
    var option: String? = null,
    var optionPrefix: String? = null,
    var commandPrev: String? = null,
    var commandNext: String? = null,
    var description: String? = null,
    var icon: String? = null,
    var execCount: Int = 0,
    var lastExecDate: LocalDateTime? = null,
) {

    constructor(entity: Link): this(
        title        = entity.title,
        group        = entity.group,
        path         = entity.path,
        relativePath = entity.relativePath,
        showConsole  = entity.showConsole,
        option       = entity.argument,
        optionPrefix = entity.commandPrefix,
        commandPrev  = entity.commandPrev,
        commandNext  = entity.commandNext,
        description  = entity.description,
        icon         = entity.icon.encodeBase64(),
        execCount    = entity.executeCount,
        lastExecDate = entity.lastExecDate,
    )

    fun toLink(): Link {
        val it = this
        return Link().apply {
            title         = it.title
            group         = it.group
            path          = toNewParameter(it.path)
            relativePath  = it.relativePath
            showConsole   = it.showConsole
            argument      = toNewParameter(it.option)
            commandPrefix = toNewParameter(it.optionPrefix)
            commandPrev   = toNewParameter(it.commandPrev)
            commandNext   = toNewParameter(it.commandNext)
            description   = it.description
            icon          = it.icon.decodeBase64()
            executeCount  = it.execCount
            lastExecDate  = it.lastExecDate
            generateKeyword()
        }
    }

    private fun toNewParameter(string: String?): String? {
        val convert = mapOf(
            "filepath" to "path",
            "dir" to "dir",
            "filename" to "file",
            "name" to "name",
            "ext" to "ext",
            "home" to "home",
        )
        var text = string
        convert.forEach{ old, new ->
            text = text?.replace("#{$old}", "\${$new}" )
        }
        return text
    }

}